### ISSUE TITLE
Activity engagment tracking for break interruptions

### ISSUE THREAD

**Issue by dilomo** [rcaelers/workrave#199]
Sometimes when you are doing something very intensely you resits taking a break. 
I was wondering if there could be some king of user activity monitor (that comply with user privacy) that proposes a break in the least active moment near the break time. Thus the users will be more willing to not postpone or skip the break.

**Comment by xerus2000** [comment]
I had a similar problem. As I found out in my recent use, it is wise to have a short non-postponeable break to rule out that resistance - feel free to read #202 :)

**Comment by dilomo** [comment]
It is different than forcing yourself with locking your computer. That's exactly what I was  trying to avoid and find the most favorable moment for beak prompt.

**Comment by rcaelers** [comment]
When it is time to take a break, Workrave will show a gentle break reminder window that does not block keyboard/mouse, does not steal the focus and disappears when you ignore it and keep working. After a number of times of ignoring the break reminder, Workrave will force you take a break. But you can disable this behavior. (When I'm very 

Workrave could start a break earlier if you are idle. But I'm not sure how to detect the 'least active moment'. By checking number of key presses per minute? Least active is probably determinted by how concentrated you are working. And Workrave cannot detect that.

Any suggestion on how to implement your idea?

**Comment by xerus2000** [comment]
Well, it already recognizes when you don't do any inputs for a few seconds and stops the times. it could be configured to automatically initiate a break then if one is due or close enough.

**Comment by dilomo** [comment]
> Workrave could start a break earlier if you are idle. But I'm not sure how to detect the 'least active moment'. By checking number of key presses per minute? Least active is probably determinted by how concentrated you are working. And Workrave cannot detect that.

Yeah its a bit tricky to implement but if we can detect the mouse movement (speed, acceleration and clicks as well as the keyboard activity) it can work. For example if I'm working in a CAD software I will move the mouse faster and without any breaks or pauses and press keys also in the same "rhythm" as the mouse movement. That would mean I'm active now. If the "rhythm" breaks or slows down then I lost focus and its a good time for break. 

Respectively for coding person, the keyboard would be a better indicator probably with some mouse activity. I can even think for profiles for the activity monitor optimized for certain profession if one approach does not work. And people can participate in collecting data for these profiles in a beta program where they label each n min and what activity level they had according to them. 

Thanks and regards,

**Comment by xerus2000** [comment]
> At the very least there should be an option that allows for idle time to be subtracted from main brake. Say i went to eat something for 6 min and then I sit down and the computer says I should have a break for 10 min. "But I just sat down." is my thought.

My approach to this is to make the forced pause length shorter (e.g. 3 minutes) - thus, when you already took a break for that time, the countdown will reset. However when it does pop up reminding you to take a break, you can still take a longer one ;)

**Comment by xerus2000** [comment]
> Yeah its a bit tricky to implement but if we can detect the mouse movement (speed, acceleration and clicks as well as the keyboard activity) it can work. For example if I'm working in a CAD software I will move the mouse faster and without any breaks or pauses and press keys also in the same "rhythm" as the mouse movement. That would mean I'm active now. If the "rhythm" breaks or slows down then I lost focus and its a good time for break.
> 
> Respectively for coding person, the keyboard would be a better indicator probably with some mouse activity. I can even think for profiles for the activity monitor optimized for certain profession if one approach does not work. And people can participate in collecting data for these profiles in a beta program where they label each n min and what activity level they had according to them.

I think this is overthinking it - as I already said, workrave notices when there is no input for a few seconds and pauses its timers, simply use that. Something like keypresses per minute seems a very dubious indicator to me. 
If you are focused thinking about a problem (so there's no input for a bit), you might as well step away from the computer for a few seconds anyways ;)

**Comment by dilomo** [comment]
> workrave notices when there is no input for a few seconds and pauses its timers, simply use that

Unfortunately, it does not work for me as it does not detect how long I was absent and breaks occur more often than they should if the program did consider my absence. I even tried different apps but I wrote here because I think there could be other people that might need that feature. 

I agree its a bit difficult to judge, but if you are focused thinking then the activity of keyboard and mouse would decrease, no? Then its OK to consider this a good time to offer a break and you can continue thinking somewhere else for 1-3-5min.

**Comment by Chealer** [comment]
@dilomo wrote:
> At the very least there should be an option that allows for idle time to be subtracted from main brake. Say i went to eat something for 6 min and then I sit down and the computer says I should have a break for 10 min. "But I just sat down." is my thought.

There is legitimacy in that, but that is a complex issue with little relation to the issue this ticket tracks, which is already a complex one. Please have that discussion in another ticket (I don't see an existing ticket about that issue).

**Comment by Chealer** [comment]
@dilomo wrote:
> That would mean I'm active now. If the "rhythm" breaks or slows down then I lost focus and its a good time for break.

It does not mean you lost focus. It can mean you are analyzing. I do many activities which require analysis, “decision-making”, and then “entering the decisions”, such as ticket triaging. For example, I spent between 20 minutes and 2 hours analyzing this ticket, yet Workrave prompted me to take a micro-break while typing this very sentence (it also prompted me multiple times during analysis). But yes, it is more efficient to take a break during analysis than while typing a reply, which might cause you to come back to an incomplete sentence having forgot how it is supposed to end.

This ticket’s topic is hard to understand from its title. Moreover, as this is an ITS, each ticket should emphasize the problem, in particular in the title. Please indicate at least 1 affected version and retitle with a short description of the issue. I suggest:
> Prompts for breaks even during most active times (offer postponing until after input peak)

[Bugzilla #450](https://workrave.org/bugzilla/450/) documents a particularly problematic consequence. That was surely a bug which was hopefully fixed since, but it does point to a starting point: try waiting until all keys/buttons are released.
This issue prompted the development of [an alternative utility, Sane Break](https://allanchain.github.io/sane-break/). It does have advantages compared to Workrave, but I found it far from Workrave when I evaluated it, in February 2025. There were 3 major releases since though, and I did not spend enough time testing, so you may want to evaluate it yourself.

[🅭🄍](https://www.philippecloutier.com/Common+infrastructure+licensing#its)

**Comment by dilomo** [comment]
I was not expecting an answer 6 years later, but thank you for engaging and commenting here. Now, an issue to be resolved by AI and pattern analysis that is typical for you, as you said. For me, it's gonna be one trigger, and for you it's gonna be another trigger for a break. 

We will have a title as you suggested. Let's discuss how to best achieve this.
